### PR TITLE
Add I/O primitives for Bigarrays

### PR DESCRIPTION
**Pull request by nojb**
Following discussion in #12360 this PR proposes adding I/O primitives for bigarrays to the standard library and Unix. I took the liberty of copying the `In_channel` and `Out_channel` functions from #12360. On top of that, `Unix` variants are also added here:

- `In_channel.input_bigarray`, `In_channel.really_input_bigarray`
- `Out_channel.output_bigarray`
- `Unix.read_bigarray`, `Unix.write_bigarray`, `Unix.single_write_bigarray`

In each case the signature of the functions are identical to the existing ones on `bytes`, except that they use `_ Bigarray.Genarray.t` in its place. Offset and length parameters are interpreted in terms of _bytes._

An alternative would be to use `_ Bigarray.Array1.t` instead of `Bigarray.Genarray.t` and interpret offset and length in terms of _elements_. I was going to do this originally, but I sensed a small difficulty with the `Unix.single_write` operation, which could fail to write a "full" element of a bigarray (but perhaps this is not a problem). Opinions welcome.

There is some code duplication in the `Unix` bindings but it is a bit difficult to share code as we take advantage of the fact that the data part of a bigarray does not move in memory to avoid copying data to an intermediate buffer and to release the runtime lock more liberally.

### CODE REVIEW

**Comment by xavierleroy** [comment]
Thanks for getting the ball rolling on this idea.  Some high-level comments following a quick look:
- The C stub code looks fine.
- The proposed API is appropriate for 1D bigarrays of characters (and 8-bit integers?) but super-confusing for other bigarrays, where the notions of byte offset and byte length don't make obvious sense.
- For arbitrary bigarrays, I'm afraid the only operations that make intuitive sense is writing the whole bigarray or reading the whole bigarray (a la really_input, raising end-of-file if too short).

A compromise could be to have "read whole" and "write whole" operations for arbitrary bigarrays in the stdlib, and byte-oriented operations over 1D char bigarrays in the Unix module... but this is to be discussed more.

**Comment by stedolan** [comment]
I think it would be nice to have functions for bigarray I/O in the stdlib (since bigarrays are in the stdlib now), but there is a reason that my original patch with @shindere was limited to `char` bigarrays: if you do I/O on non-`char` bigarrays, you have to think about endianness, partial writes, offsets and lengths that are non-integer multiples of the element size, and so on.

This is because channels read and write byte sequences, so some conversion needs to be done if you are reading and writing sequences of things other than bytes. I'm not sure that the stdlib I/O functions are the right place for such a conversion API: perhaps the user should do the conversion to bytes using functions from Bigarray, and the I/O should operate only on bigarrays of bytes?

**Comment by nojb** [comment]
> perhaps the user should do the conversion to bytes using functions from Bigarray, and the I/O should operate only on bigarrays of bytes?

This sounds reasonable to me.

**Comment by nojb** [comment]
> > perhaps the user should do the conversion to bytes using functions from Bigarray, and the I/O should operate only on bigarrays of bytes?
> 
> This sounds reasonable to me.

I am planning to rework the PR to restrict I/O primitives to 1-D char bigarrays, with the following signatures:

```ocaml
Unix.read_bigarray : Unix.file_descr -> (char, int8_unsigned_int, _) Bigarray.Array1.t -> int -> int -> int
Unix.write_bigarray : Unix.file_descr -> (char, int8_unsigned_int, _) Bigarray.Array1.t -> int -> int -> int
Unix.single_write_bigarray : Unix.file_descr -> (char, int8_unsigned_int, _) Bigarray.Array1.t -> int -> int -> int
In_channel.input_bigarray : t -> (char, int8_unsigned_int, _) Bigarray.Array1.t -> int -> int -> int
In_channel.really_input_bigarray : t -> (char, int8_unsigned_int, _) Bigarray.Array1.t -> int -> int -> unit
Out_channel.output_bigarray : t -> (char, int8_unsigned_int, _) Bigarray.Array1.t -> int -> int -> unit
```

Please speak up if you have any objections!

**Comment by yallop** [comment]
I think it'd be clearer to also restrict the layout to `c_layout`, to avoid any possible confusion around 0-based and 1-based indexing.  An alternative would be to carefully document the behaviour on Fortran-layout bigarrays, e.g. whether passing `~pos:1` to `write_bigarray` will start writing at the first or the second element.

**Comment by nojb** [comment]
> I think it'd be clearer to also restrict the layout to `c_layout`, to avoid any possible confusion around 0-based and 1-based indexing.

Good point, will do.

**Comment by xavierleroy** [comment]
> I am planning to rework the PR to restrict I/O primitives to 1-D char bigarrays, with the following signatures: `Unix.read_bigarray : Unix.file_descr -> (char, int8_unsigned_int, _) Bigarray.Array1.t -> int -> int -> int`

I agree that the layout should better be forced  to be `c_layout`, to avoid misunderstandings w.r.t. Fortran layout.  On the other hand, the first parameter `char` could be generalized to `_`, so as to support both `char` and `int8_unsigned` bigarray kinds.  Isn't that cute?

**Comment by xavierleroy** [comment]
> This is because channels read and write byte sequences, so some conversion needs to be done if you are reading and writing sequences of things other than bytes. I'm not sure that the stdlib I/O functions are the right place for such a conversion API: perhaps the user should do the conversion to bytes using functions from Bigarray, and the I/O should operate only on bigarrays of bytes?

On the one hand, I agree that endianness differences are best handled by using marshaling (input_value/output_value).

On the other hand, the memory-mapping API (`Unix.map_file`) supports all kinds of bigarrays without any conversions, just using the platform's native endianness.  For I/O on pipes or sockets, memory mapping is not an option, hence functions for reading and writing bigarrays without conversions could make sense.

On the third hand, just like we have `Bigarray.reshape*` functions to change the dimensions of bigarrays without copying, we could have `Bigarray.repr*` functions that provide a view of a bigarray as an `int8_unsigned` 1D bigarray, over which I/O can be performed.  Looks like the most flexible approach.

File: otherlibs/unix/write_unix.c
Comment (yallop) [inline otherlibs/unix/write_unix.c:71]: This is not GC-safe: the bigarray could be collected after `caml_enter_blocking_section`, making `buf` invalid, e.g. in code like this:

```ocaml
Unix.write_bigarray fd (Bigarray.Array1.init ...) pos len
```

It should be sufficient to guard `vbuf` with one of the `CAMLparam*` macros, since the `data` itself won't be moved as long as the underlying bigarray is alive.

The other functions have the same issue (even if it's unlikely to arise in practice for reads).

**Review by yallop — CHANGES_REQUESTED** [review]

**Comment by nojb** [comment]
> On the other hand, the first parameter `char` could be generalized to `_`, so as to support both `char` and `int8_unsigned` bigarray kinds. Isn't that cute?

It is. Amended!

File: otherlibs/unix/write_unix.c
Comment (nojb) [reply otherlibs/unix/write_unix.c:71]: Oops, good catch! It looks like the same bug was also present in the handling of Windows file descriptors (which are custom blocks). I believe both are now fixed. Thanks!

File: otherlibs/unix/read_unix.c
Comment (gasche) [inline otherlibs/unix/read_unix.c]: Color me confused, isn't this change *introducing* a GC safety bug? Once you enter a blocking section, another thread may perform a GC and move `vbuf` under your feet. See for example

https://github.com/ocaml/ocaml/blob/5ff9ea2780bcae5c1bbc675c9e2f341e080cba42/otherlibs/unix/lseek_unix.c#L57-L63

File: otherlibs/unix/write_unix.c
Comment (gasche) [reply otherlibs/unix/write_unix.c:71]: Ah, I had missed this comment which was the context for the change in the new commit. I think that rooting `vbuf` is indeed necessary, but you should still access its fields before entering the blocking section.

File: otherlibs/unix/read_unix.c
Comment (nojb) [reply otherlibs/unix/read_unix.c]: > Color me confused, isn't this change _introducing_ a GC safety bug? Once you enter a blocking section, another thread may perform a GC and move `vbuf` under your feet.

I was convinced that registering `vbuf` as a root was enough to guard against that, but now am not so sure anymore.

File: otherlibs/unix/write_unix.c
Comment (nojb) [reply otherlibs/unix/write_unix.c:71]: OK. But just for my own understanding: do you agree that the present code (rooting `vbuf` + accesing the fields _inside_ the blocking section) is safe as it is?

Comment (gasche) [reply otherlibs/unix/write_unix.c:71]: I think that accessing arguments of `vbuf` inside the blocking section is *not* correct. My understanding is that you need to have the runtime lock to read inside a memory block owned by the GC, otherwise you may race with a GC moving the value around. (This is consistent with the comment in lseek_unix that is quoted in the other thread.)

Comment (nojb) [reply otherlibs/unix/write_unix.c:71]: Right, makes sense. I went back to accessing the blocks outside the runtime block. Hopefully the code is correct now.

Comment (gasche) [reply otherlibs/unix/write_unix.c:71]: I believe that the code is correct. This said, I am not sure why you are not following the "full" discipline of registering all the value parameters (even those that we assume are immediate). It would make the code easier to read, and I would bet that there is no noticeable performance differrence.

File: runtime/io.c
Comment (gasche) [inline runtime/io.c:876]: You probably also need to register `vchannel` as a root here.

File: otherlibs/unix/write_unix.c
Comment (nojb) [reply otherlibs/unix/write_unix.c:71]: > It would make the code easier to read,

Done. Thanks.

File: runtime/io.c
Comment (nojb) [reply runtime/io.c:876]: Done. Thanks.

File: otherlibs/unix/unixsupport.h
Comment (stedolan) [inline otherlibs/unix/unixsupport.h]: Since the result of this points inside the GC object `v`, the resulting value is only valid until the next GC point and is entirely invalid inside a blocking section.

File: otherlibs/unix/write_unix.c
Comment (stedolan) [reply otherlibs/unix/write_unix.c:71]: @gasche is correct with this description:
> I think that accessing arguments of vbuf inside the blocking section is not correct. My understanding is that you need to have the runtime lock to read inside a memory block owned by the GC, otherwise you may race with a GC moving the value around.

The issue is fixed here, but still present in `read_win32.c` and `write_win32.c` - since `Descr_val(vfd)` points into GC-managed memory, it cannot be used inside a blocking section.

Comment (nojb) [reply otherlibs/unix/write_unix.c:71]: Of course! Pushed a fix for the Win32 stubs.

**Comment by stedolan** [comment]
The latest GC safety patch looks good to me, although appveyor points out that one remaining `fd->kind` should become a `Descr_kind_val(fd)` at `read_win32.c:75`.

**Comment by stedolan** [comment]
> we could have Bigarray.repr* functions that provide a view of a bigarray as an int8_unsigned 1D bigarray

I agree with the third hand here - the `Bigarray.repr*` functions sound useful, and I actually thought we already had them! (For a separate PR, though)

**Comment by xavierleroy** [comment]
> the Bigarray.repr* functions sound useful, and I actually thought we already had them! 

We have `reshape` functions that change the number and values of dimensions, but not the element kind.  `repr` functions would serve a different purpose, but I agree they would be useful too.

**Comment by nojb** [comment]
> The latest GC safety patch looks good to me, although appveyor points out that one remaining `fd->kind` should become a `Descr_kind_val(fd)` at `read_win32.c:75`.

Thanks, fixed!

File: otherlibs/unix/write_win32.c
Comment (stedolan) [inline otherlibs/unix/write_win32.c]: I think this line shouldn't be here, and is causing the remaining appveyor failure.

File: otherlibs/unix/read_unix.c
Comment (dra27) [inline otherlibs/unix/read_unix.c:43]: Up to you, but there's no need to register things that are required to be integers with the GC:
```suggestion
  CAMLparam2(fd, vbuf);
```
(equally, that thinking is how the `fd` wasn't registered as a root...)

Comment (yallop) [reply otherlibs/unix/read_unix.c:43]: See https://github.com/ocaml/ocaml/pull/12365#discussion_r1263491872

File: otherlibs/unix/write_win32.c
Comment (nojb) [reply otherlibs/unix/write_win32.c]: Indeed, fixed. Thanks!

**Comment by nojb** [comment]
This PR needs two official approvals if it is to move forward. Any takers?

Thanks!

**Comment by shindere** [comment]
Nicolás Ojeda Bär (2023/07/27 05:41 -0700):
> This PR needs two official approvals if it is to move forward. Any
> takers?

Sure. Would you please mind squashing all the commits?

**Comment by nojb** [comment]
> Sure. Would you please mind squashing all the commits?

Sure, done.

**Comment by shindere** [comment]
Thanks! What's the status of the change requested by @yallop? GH shows it
to me as not taken into account, is that correct?

**Comment by nojb** [comment]
> What's the status of the change requested by @yallop? GH shows it to me as not taken into account, is that correct?

I believe all mentioned issues have been addressed.

**Review by shindere — APPROVED** [review]

**Review by xavierleroy — COMMENTED** [review]
I reviewed the implementation.   I think it's OK except for the cases where the number of bytes to read or write is not representable as an `int` or a `DWORD`.  I took the liberty to push fixes directly on this PR: one commit is for Unix, the other for Win32.  Let me know what you think.

I'm also tempted to factor out the C code between the "write" and "single_write" cases, but haven't done anything in this direction yet.

**Comment by c-cube** [comment]
This is going to be another nail in the coffin for [modular IO](https://github.com/ocaml/RFCs/blob/master/rfcs/modular_io.md), isn't it? More C functions, now operating on something that's not byte buffers…

**Comment by xavierleroy** [comment]
> I'm also tempted to factor out the C code between the "write" and "single_write" cases, but haven't done anything in this direction yet.

Here is a first try, on a personal branch:  https://github.com/ocaml/ocaml/commit/8b954fb57572e10e57ade1d098b560c2934276b1

**Comment by nojb** [comment]
> I reviewed the implementation. I think it's OK except for the cases where the number of bytes to read or write is not representable as an `int` or a `DWORD`. I took the liberty to push fixes directly on this PR: one commit is for Unix, the other for Win32. Let me know what you think.

Thanks for the fix and the careful comments. Both commits look good to me.

**Comment by nojb** [comment]
> > I'm also tempted to factor out the C code between the "write" and "single_write" cases, but haven't done anything in this direction yet.
> 
> Here is a first try, on a personal branch: [8b954fb](https://github.com/ocaml/ocaml/commit/8b954fb57572e10e57ade1d098b560c2934276b1)

Thanks, looks good to me so I cherry-picked to this branch. Should we do the same for `Unix.write` and `Unix.single_write`?

**Comment by xavierleroy** [comment]
> Should we do the same for Unix.write and Unix.single_write?

I thought about it, but was afraid to break 3rd-party reimplementations of the Unix module (JS_of_ocaml, maybe?) that might assume there are two different primitives.  So, let's leave it as that.

**Review by xavierleroy — APPROVED** [review]
Looks good to me, approving!  Formally, a second approval is needed, as this is a stdlib extension.

**Comment by shindere** [comment]
Xavier Leroy (2023/08/18 10:43 -0700):
> @xavierleroy approved this pull request.
>
> Looks good to me, approving!  Formally, a second approval is needed,
> as this is a stdlib extension.

Isn't your approvla the second one?

I did approve thisPR a few weeks ago but perhaps you'd prefer the second
approval to be from a core dev more seasoned with stdlib changes?

**Comment by xavierleroy** [comment]
> I did approve thisPR a few weeks ago

Ah, sorry, I forgot about it (was one month ago) and didn't look at the full history.

> but perhaps you'd prefer the second approval to be from a core dev more seasoned with stdlib changes?

I'm neutral.  More eyeballs is always good, but as stdlib extensions go. this PR isn't controversial, I believe.

At any rate, I'll look into this again when I'm back next week.

**Comment by shindere** [comment]
Xavier Leroy (2023/08/22 08:59 -0700):
> Ah, sorry, I forgot about it (was one month ago) and didn't look at
> the full history.

No problem. :)

> > but perhaps you'd prefer the second approval to be from a core dev more seasoned with stdlib changes?
>
> I'm neutral.  More eyeballs is always good,

Yeah, especially given the poor quality of mine. Sorry, coulnd't resist.

> but as stdlib extensions go. this PR isn't controversial, I believe.

I don't believe either. :)

> At any rate, I'll look into this again when I'm back next week.

Thanks. Its merge will then unblock #12360 which will be made simpler
once rebased, I expect.

**Comment by nojb** [comment]
> > but perhaps you'd prefer the second approval to be from a core dev more seasoned with stdlib changes?
> 
> I'm neutral. More eyeballs is always good, but as stdlib extensions go. this PR isn't controversial, I believe.

@shindere's review took place before the changes explained in https://github.com/ocaml/ocaml/pull/12365#pullrequestreview-1569586449 so it would be best to have another review of the current state of the patch.

> At any rate, I'll look into this again when I'm back next week.

Thanks!

File: stdlib/out_channel.mli
Comment (dra27) [inline stdlib/out_channel.mli]: Typo:
```suggestion
(** Same as {!output} but take the data from a bigarray.
```

File: stdlib/in_channel.mli
Comment (dra27) [inline stdlib/in_channel.mli]: Consistency with `Unix.read_bigarray`'s description?
```suggestion
(** Same as {!input}, but read the data into a bigarray.
```

File: stdlib/in_channel.mli
Comment (dra27) [inline stdlib/in_channel.mli]: Similarly?
```suggestion
(** Same as {!really_input}, but read the data into a bigarray.
```

**Review by dra27 — APPROVED** [review]
LVGTM2! There is one actual typo to fix in the doc comment for `Out_channel.output_bigarray`.

Only because of that, the description of `Unix.read_bigarray` uses the verb "read" in its description yet the descriptions of `In_channel.input_bigarray` and `In_channel.really_input_bigarray` use the verb "write" which read oddly in this unusual position of reading them all at once. I'd alter the `In_channel` functions to use "read the data into a bigarray" instead.

Finally, in the doc strings, there's inconsistency between "take data" and "take the data" between otherwise identical descriptions - FWIW I'd go for adding _the_ (i.e. "read _the_ data into a bigarray" and "take the data from a bigarray").

**Comment by shindere** [comment]
Cool! Thanks a lot for the review!

Perhaps worth taking the opportunity of dong the fixes to squash all the
commits or at least make sure their number is minimal and the history
coherent?

**Comment by nojb** [comment]
> Perhaps worth taking the opportunity of dong the fixes to squash all the commits or at least make sure their number is minimal and the history coherent?

No need, I'll squash the PR when merging.

**Comment by nojb** [comment]
> LVGTM2!

Thanks @dra27 for your review! I accepted all your suggestions.

I suggest we wait for @xavierleroy's second look before merging.

**Comment by xavierleroy** [comment]
I had a (quick) second look and I think it's high time to merge this PR!  Thanks to all who participated.

**Comment by nojb** [comment]
Thanks!

**Pull request merged**
